src/python/dxpy/sugar/context_handlers.py

```diff
@@ -0,0 +1,149 @@
+from __future__ import print_function
+import os
+import sys
+import tempfile
+import subprocess
+import contextlib
+import shutil
+
+
+@contextlib.contextmanager
+def set_env(**environ):
+    """
+    Context manager generator to temporarily set the subprocess environment variables.
+
+    Args:
+        environ (dict): Environment variable to set
+
+    Yields:
+        An environment with environment variables set as specified.
+        On exit, the environment will return to previous configuration.
+
+    Examples:
+        Usage 1: Set environment variable
+        # inside environment
+        >>> with set_env(PLUGINS_DIR=u'test/plugins'):
+        ...    "PLUGINS_DIR" in os.environ
+        True
+
+        # outside environment
+        >>> "PLUGINS_DIR" in os.environ
+        False
+
+        Usage 2: Unset environment variable
+        >>> with set_env(PYTHONPATH=''):
+        ...    print(os.environ["PYTHONPATH"])
+        <BLANKLINE>
+
+        Usage 3: Manipulate multiple variables
+        >>> myenv = {"PLUGINS_DIR": u'test/plugins', "PYTHONPATH": u'some/python/path'}
+        >>> with set_env(**myenv):
+        ...   print(os.environ["PLUGINS_DIR"])
+        ...   print(os.environ["PYTHONPATH"])
+        test/plugins
+        some/python/path
+    """
+    old_environ = dict(os.environ)
+    os.environ.update(environ)
+    try:
+        yield
+    finally:
+        os.environ.clear()
+        os.environ.update(old_environ)
+
+@contextlib.contextmanager
+def cd(new_path=None, temp_dir=None):
+    """Context manager that changes the current working directory to `new_path`,
```

Comment:
I don't think temp_dir arg is needed. You could just have it work such that:
* If no argument is specified, create a temp dir (using tempfile.mkdtemp), cd to it and cd back and delete the directory before returning
* If an argument is specified and the directory doesn't exist, create it, cd to it, and cd back and delete the directory before returning
* If an argument is specified and the directory does exist, cd to the directory, cd to it, and cd back but don't delete the directory before returning.

You could have 'create' and 'cleanup' kwargs for the user to override the default behavior (if for example they don't want the tempdir to be deleted).

Reply:
I think "create" is redundant since a non-existent path would always have to be created, where as an existent path should be cd'd to without overwriting it. A "cleanup" command makes sense. We could replace `new_path` and `temp_dir` with just a `target_path` and `delete_on_exit` options. The only concern is user error if they cd to a root directory and then enable `delete_on_exit`.

Reply:
That is why I don't like the delete_on_exit arg. IMO the context manager should only be able to delete a directory that it creates itself.

Reply:
That makes sense to me.